core-bundle/src/EventListener/DataContainer/PageUrlListener.php

```diff
@@ -191,65 +199,85 @@ private function recursiveValidatePages(int $pid, PageModel $rootPage): void
         }
 
-        $pageAdapter = $this->framework->getAdapter(PageModel::class);
-        $currentUrl = $this->buildUrl($currentAlias, $currentPrefix, $currentSuffix);
+        $aliasPages = $this->framework->getAdapter(PageModel::class)->findSimilarByAlias($currentPage);
 
-        foreach ($aliasIds as $aliasId) {
-            $aliasPage = $pageAdapter->findWithDetails($aliasId);
+        if (null === $aliasPages) {
+            return false;
+        }
 
-            if (null === $aliasPage) {
-                continue;
-            }
+        $routeCollection = new RouteCollection();
 
+        foreach ($aliasPages as $aliasPage) {
             if (!$this->pageRegistry->isRoutable($aliasPage)) {
                 continue;
             }
 
-            $aliasDomain = $aliasPage->domain;
-            $aliasPrefix = $aliasPage->urlPrefix;
-            $aliasSuffix = $aliasPage->urlSuffix;
-
+            // If page has the same root, inherit root page settings from post data
             if ($currentPage->rootId === $aliasPage->rootId) {
-                $aliasDomain = $currentDomain;
-                $aliasPrefix = $currentPrefix;
-                $aliasSuffix = $currentSuffix;
+                $aliasPage->domain = $currentPage->domain;
+                $aliasPage->urlPrefix = $currentPage->urlPrefix;
+                $aliasPage->urlSuffix = $currentPage->urlSuffix;
             }
 
-            $aliasUrl = $this->buildUrl($aliasPage->alias, $aliasPrefix, $aliasSuffix);
+            $aliasRoute = $this->pageRegistry->getRoute($aliasPage);
+
+            // Even if we cannot generate the path because of parameter requirements,
+            // two pages can never have the same path AND the same requirements. This
+            // could be two regular pages with same alias and  `requireItem` enabled.
+            if (
+                null === $currentUrl
+                && $currentRoute->getPath() === $aliasRoute->getPath()
+                && 0 === ($currentRoute->getRequirements() <=> $aliasRoute->getRequirements())
+            ) {
+                if ($throw) {
+                    throw new DuplicateAliasException($currentRoute->getPath());
+                }
 
-            if ($currentDomain !== $aliasDomain || $currentUrl !== $aliasUrl) {
-                continue;
+                return true;
             }
 
-            // Duplicate alias found
-            if ($throw) {
-                throw new DuplicateAliasException($currentUrl);
-            }
+            $routeCollection->add('tl_page.'.$aliasPage->id, $aliasRoute);
+        }
 
-            return true;
+        if (null === $currentUrl || 0 === $routeCollection->count()) {
+            return false;
         }
 
-        return false;
+        $request = Request::create($currentUrl);
+
+        try {
+            $this->routeMatcher->finalMatch($routeCollection, $request);
+            // TODO return value is the which page conflicts, should we display that?
```

Comment:
What about this comment?

Reply:
I suggest we merge it as-is and add a new issue to discuss this (maybe in the next call).

Reply:
https://github.com/contao/contao/issues/3903